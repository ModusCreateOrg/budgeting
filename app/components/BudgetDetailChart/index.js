// @flow

import * as React from 'react';
import { scaleOrdinal } from 'd3';

import isPositive from '../../utils/isPositive';
import getPercentage from '../../utils/getPercentage';
import DonutChart from '../DonutChart';

type BudgetDetailChartProps = {
  amount: Number,
  totalInflow: Number,
  totalOutflow: Number,
  description: String,
};

const BudgetDetailChart = ({ amount, totalInflow, totalOutflow, description }: BudgetDetailChartProps) => {
  const isInflow = isPositive(amount);
  const percentage = isInflow ? getPercentage(amount, totalInflow) : getPercentage(amount, totalOutflow);
  const chartData = [
    {
      description,
      key: 0,
      value: percentage,
    },
    {
      description: isInflow ? 'Other Inflows' : 'Other Outflows',
      key: 1,
      value: Number((100 - percentage).toFixed(2)),
    },
  ];
  return (
    <DonutChart
      dataLabel={'description'}
      dataKey={'key'}
      data={chartData}
      formatLabel={false}
      color={scaleOrdinal(['#edc', '#ccc', '#ccc'])}
    />
  );
};

BudgetDetailChart.defaultProps = {
  amount: 1,
  totalInflow: 0,
  totalOutflow: 0,
  description: '',
};

export default BudgetDetailChart;
