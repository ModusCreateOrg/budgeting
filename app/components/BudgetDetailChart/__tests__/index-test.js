import React from 'react';
import BudgetDetailChart from 'components/BudgetDetailChart';
import { shallow } from 'enzyme';

describe('BudgetDetailCard', () => {
  const defaultProps = {
    amount: 20,
    totalInflow: 100,
    totalOutflow: -200,
    description: 'test',
  };
  const createWrapper = overridenProps => {
    const props = {
      ...defaultProps,
      ...overridenProps,
    };
    return shallow(<BudgetDetailChart {...props} />);
  };
  it('should renders correctly', () => {
    const wrapper = createWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('should calculate the correct percentage for inflows', () => {
    const wrapper = createWrapper();
    const chart = wrapper.find('DonutChart').at(0);
    expect(chart.prop('data')).toEqual([
      {
        description: 'test',
        key: 0,
        value: 20,
      },
      {
        description: 'Other Inflows',
        key: 1,
        value: 80,
      },
    ]);
  });

  it('should calculate the correct percentage for outflows', () => {
    const wrapper = createWrapper({ amount: -20 });
    const chart = wrapper.find('DonutChart').at(0);
    expect(chart.prop('data')).toEqual([
      {
        description: 'test',
        key: 0,
        value: 10,
      },
      {
        description: 'Other Outflows',
        key: 1,
        value: 90,
      },
    ]);
  });
});
