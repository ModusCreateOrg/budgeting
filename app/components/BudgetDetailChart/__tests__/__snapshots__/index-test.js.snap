// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetDetailCard should renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BudgetDetailChart
    amount={20}
    description="test"
    totalInflow={100}
    totalOutflow={-200}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "color": [Function],
      "data": Array [
        Object {
          "description": "test",
          "key": 0,
          "value": 20,
        },
        Object {
          "description": "Other Inflows",
          "key": 1,
          "value": 80,
        },
      ],
      "dataKey": "key",
      "dataLabel": "description",
      "dataValue": "value",
      "formatLabel": false,
      "height": 300,
      "innerRatio": 4,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "color": [Function],
        "data": Array [
          Object {
            "description": "test",
            "key": 0,
            "value": 20,
          },
          Object {
            "description": "Other Inflows",
            "key": 1,
            "value": 80,
          },
        ],
        "dataKey": "key",
        "dataLabel": "description",
        "dataValue": "value",
        "formatLabel": false,
        "height": 300,
        "innerRatio": 4,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
